feat: Implement Task Commenting Feature

ShubhamAggarwal0812 opened the pull request:
# Pull Request for Task Commenting Feature

## Description
This PR introduces the task commenting feature to our application. Users can now add, edit, delete, and view comments on tasks. The implementation includes both backend and frontend changes to support this functionality.

## Why is this change needed?
This change is necessary to allow users to communicate and collaborate on tasks by adding comments. It enhances the task management system by enabling better communication and feedback.

## Changes Overview
### Backend
1. **New Endpoints:**
   - `POST /comments`: Create a new comment for a task.
   - `GET /comments/:taskId`: Fetch all comments for a specific task.
   - `PUT /comments/:commentId`: Update a specific comment.
   - `DELETE /comments/:commentId`: Delete a specific comment.

2. **Database Models:**
   - **CommentDB Schema:** Defines the structure of comments stored in the database.

3. **Controllers:**
   - **CommentController:** Handles the HTTP requests for comments.

4. **Services:**
   - **CommentService:** Contains the business logic for managing comments.

5. **Utils:**
   - **CommentUtil:** Converts database objects to API responses.

### Frontend
1. **Components:**
   - **CommentList:** Displays a list of comments for a task.
   - **AddComment:** Allows users to add a new comment to a task.

2. **Context:**
   - **CommentContext:** Manages the state and actions related to comments.

3. **Services:**
   - **CommentService:** Handles API calls related to comments.

4. **Types:**
   - **Comment:** Defines the structure of comment objects used in the frontend.

## API Changes
- **POST /comments**
  - **Request Body:**
    ```json
    {
      "taskId": "string",
      "comment": "string"
    }
    ```
  - **Response:**
    ```json
    {
      "id": "string",
      "task": "string",
      "account": {
        "id": "string",
        "firstName": "string",
        "lastName": "string",
        "username": "string"
      },
      "comment": "string",
      "createdAt": "string",
      "updatedAt": "string"
    }
    ```

- **GET /comments/:taskId**
  - **Response:**
    ```json
    [
      {
        "id": "string",
        "task": "string",
        "account": {
          "id": "string",
          "firstName": "string",
          "lastName": "string",
          "username": "string"
        },
        "comment": "string",
        "createdAt": "string",
        "updatedAt": "string"
      }
    ]
    ```

- **PUT /comments/:commentId**
  - **Request Body:**
    ```json
    {
      "taskId": "string",
      "comment": "string"
    }
    ```
  - **Response:**
    ```json
    {
      "id": "string",
      "task": "string",
      "account": {
        "id": "string",
        "firstName": "string",
        "lastName": "string",
        "username": "string"
      },
      "comment": "string",
      "createdAt": "string",
      "updatedAt": "string"
    }
    ```

- **DELETE /comments/:commentId**
  - **Response:** 204 No Content

## Database Schema Changes
No changes were made to the existing database schema. The new `CommentDB` schema was added to support comments.

## Tests
### Automated Test Cases Added: (Yet to add)
1. **Creating a Comment:** Validates that a comment can be created successfully.
2. **Fetching Comments for a Task:** Ensures that comments for a task are fetched correctly.
3. **Updating a Comment:** Checks that a comment can be updated.
4. **Deleting a Comment:** Confirms that a comment can be deleted.

### Manual Test Cases Run:
1. **Creating a Comment:**
   - Navigate to a task.
   - Add a comment.
   - Verify that the comment appears in the list.
2. **Fetching Comments for a Task:**
   - Refresh the task page.
   - Ensure that all comments are displayed correctly.
3. **Updating a Comment:**
   - Edit an existing comment.
   - Verify that the changes are saved and displayed.
4. **Deleting a Comment:**
   - Delete a comment.
   - Confirm that the comment is removed from the list.

## Meta Information
- **Assignee:** @ShubhamAggarwal0812 
- **Labels:** feature, task-commenting, backend, frontend

## Self Review
This PR has been self-reviewed to ensure quality and consistency.

Thank you for reviewing this PR.

Commit:
feat: Implement Task Commenting Feature

aniketrathi commented on src/apps/backend/modules/comment/internal/comment-util.ts:
```diff
@@ -0,0 +1,15 @@
+    const comment = new Comment();
+    comment.id = commentDb._id.toString();
+    comment.task = commentDb.task.toString();
+    comment.account = commentDb.account;
```
`comment.account = commentDb.account` is not correct. We have to do it this way if the type of comment.account is an object ID then comment.account = comment.account.toString() else if it is an object then comment.account = AccountUtil.convertAccountDBToAccount(commentDb.account).

aniketrathi commented on src/apps/backend/modules/comment/internal/store/comment-db.ts:
```diff
@@ -0,0 +1,38 @@
+    comment: {
+      type: String,
+      required: true,
+    },
```
Please add an `active` field to support soft-deletion.

aniketrathi commented on src/apps/backend/modules/comment/internal/comment-writer.ts:
```diff
@@ -0,0 +1,69 @@
+
+    await CommentRepository.findOneAndDelete({
+      _id: comment.id,
+    });
```
Avoid hard deletion

aniketrathi commented on src/apps/backend/modules/comment/rest-api/comment-router.ts:
```diff
@@ -0,0 +1,18 @@
+    router.use(accessAuthMiddleware);
+
+    router.post('/', ctrl.createComment);
+    router.get('/:taskId', ctrl.getComments);
```
Remove `taskId` from here

aniketrathi commented on src/apps/backend/modules/comment/rest-api/comment-server.ts:
```diff
@@ -0,0 +1,12 @@
+    const { server } = this;
+    const router = new CommentRouter();
+
+    server.use('/comments', router.router);
```
`/tasks/:taskId/comments`

aniketrathi commented on src/apps/backend/modules/comment/types.ts:
```diff
@@ -0,0 +1,44 @@
+import { HttpStatusCodes } from '../http';
+
+export class Comment {
+  id!: string;
```
Remove !
why we need this?

aniketrathi commented on src/apps/backend/modules/comment/types.ts:
```diff
@@ -0,0 +1,44 @@
+export class Comment {
+  id!: string;
+  task!: string;
+  account!: any;
```
Avoid use of any

aniketrathi commented on src/apps/backend/modules/task/rest-api/task-router.ts:
```diff
@@ -15,5 +16,8 @@ export default class TaskRouter extends ApplicationRouter {
     router.delete('/:id', ctrl.deleteTask);
+
+    const commentRouter = new CommentRouter();
+    router.use('/:taskId/comments', commentRouter.router);
```
move this to comment-router

Commit:
feat: Refactor comment module, soft deletion, router, util and type

ShubhamAggarwal0812 commented:
### PR Update: Implementations Based on Feedback

I have refactored both the backend and frontend code for the commenting feature as per the feedback provided. Below are the changes made:

#### Backend
1. **Updated Comment Util:**
   - Refactored `convertCommentDBToComment` to handle both `ObjectId` and `Account` types for `comment.account`.

2. **Comment Types:**
   - Replaced the use of `any` with proper types for `comment.account`.
   - Removed the non-null assertion operator (`!`) for better type safety.

3. **Soft Deletion:**
   - Added an `active` field in the `CommentDB` schema to support soft deletion.
   - Updated the `deleteComment` method to perform soft deletion by setting `active` to `false`.

4. **Comment Router and Server:**
   - Adjusted the routing structure to use `/tasks/:taskId/comments` for all comment-related operations.
   - Ensured consistency in endpoints for creating, updating, deleting, and fetching comments.

5. **Comment Serializer:**
   - Modified the serializer to handle both `string` and `Account` types for `comment.account`.

#### Frontend
1. **Comment Provider:**
   - Updated context functions to match the new backend endpoints.
   - Adjusted the `deleteComment` function to include the `taskId` parameter.

2. **Comment List Component:**
   - Refactored to properly handle the new account structure.
   - Ensured proper display of account details in the comment section.

3. **Comment Service:**
   - Updated API service methods to match the new endpoint structure for comments under tasks.

These changes ensure proper integration between the frontend and backend while adhering to the feedback provided. The code has been tested to confirm that it works as expected.

Please review the updates and let me know if there are any further changes required.

aniketrathi commented on src/apps/backend/modules/comment/rest-api/comment-serializer.ts:
```diff
@@ -0,0 +1,18 @@
+  task: comment.task,
+  account:
+    typeof comment.account === 'string'
+      ? { id: comment.account }
```
if an account is a string then just do comment. account no need for an id field inside the object.

aniketrathi commented on src/apps/backend/modules/comment/types.ts:
```diff
@@ -0,0 +1,61 @@
+    this.comment = comment;
+    this.createdAt = createdAt;
+    this.updatedAt = updatedAt;
+  }
```
Why we need this constructor?

aniketrathi commented on src/apps/backend/modules/comment/internal/comment-util.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  private static convertAccount(
+    account: Types.ObjectId | Account,
```
convertPopulatedAccountDBToAccount(
    account: Types.ObjectId | Account,
 ): string | Account

aniketrathi commented:
Write automated test cases
